File: web/lib/engine/crawler.ts
import * as cheerio from 'cheerio'

const EMAIL_RE = /[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/g
const TIMEOUT_MS = 15_000

export interface CrawlResult {
  images: string[]
  emails: string[]
  title: string
  description: string
}

function resolveUrl(src: string, baseUrl: string): string | null {
  if (!src || src.startsWith('data:')) return null
  try {
    return new URL(src, baseUrl).href
  } catch {
    return null
  }
}

export async function crawl(url: string): Promise<CrawlResult> {
  const controller = new AbortController()
  const timer = setTimeout(() => controller.abort(), TIMEOUT_MS)
  try {
    const resp = await fetch(url, {
      signal: controller.signal,
      headers: { 'User-Agent': 'AuraGuard/1.0' },
    })
    if (!resp.ok) {
      console.warn(`Failed to fetch ${url}: ${resp.status}`)
      return { images: [], emails: [], title: '', description: '' }
    }
    const html = await resp.text()
    const $ = cheerio.load(html)

    const images: string[] = []
    $('img').each((_, el) => {
      const src = $(el).attr('src') || $(el).attr('data-src') || ''
      if (src) {
        const resolved = resolveUrl(src, url)
        if (resolved) images.push(resolved)
      }
    })

    const emails = new Set<string>()
    $('a[href]').each((_, el) => {
      const href = $(el).attr('href') || ''
      if (href.startsWith('mailto:')) {
        emails.add(href.slice(7).split('?')[0])
      }
    })
    for (const match of html.matchAll(EMAIL_RE)) {
      emails.add(match[0].toLowerCase())
    }

    const title = $('title').first().text().trim()
    const desc = $('meta[name="description"]').attr('content')
      || $('meta[property="og:description"]').attr('content')
      || ''

    console.log(`Crawled ${url} — ${images.length} images, ${emails.size} emails`)
    return { images, emails: [...emails], title, description: desc }
  } catch (err: any) {
    console.warn(`Failed to fetch ${url}: ${err.message}`)
    return { images: [], emails: [], title: '', description: '' }
  } finally {
    clearTimeout(timer)
  }
}

File: engine/scraper/crawler.py
import re
import logging
import requests
from bs4 import BeautifulSoup

logger = logging.getLogger(__name__)

EMAIL_PATTERN = re.compile(r"[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}")
TIMEOUT = 15


def crawl(url: str) -> dict:
    logger.info("Crawling %s", url)
    try:
        resp = requests.get(url, timeout=TIMEOUT, headers={"User-Agent": "AuraGuard/1.0"})
        resp.raise_for_status()
    except Exception as e:
        logger.warning("Failed to fetch %s: %s", url, e)
        return {"images": [], "emails": [], "title": "", "description": ""}

    soup = BeautifulSoup(resp.text, "html.parser")

    images = []
    for tag in soup.find_all("img"):
        src = tag.get("src") or tag.get("data-src") or ""
        if src.startswith("//"):
            src = "https:" + src
        elif src.startswith("/"):
            from urllib.parse import urlparse
            parsed = urlparse(url)
            src = f"{parsed.scheme}://{parsed.netloc}{src}"
        if src and not src.startswith("data:"):
            images.append(src)

    emails = set()
    for tag in soup.find_all("a", href=True):
        href = tag["href"]
        if href.startswith("mailto:"):
            emails.add(href[7:].split("?")[0])
    for match in EMAIL_PATTERN.findall(resp.text):
        emails.add(match.lower())

    title = ""
    t_tag = soup.find("title")
    if t_tag:
        title = t_tag.get_text(strip=True)

    description = ""
    meta = soup.find("meta", attrs={"name": "description"}) or soup.find("meta", attrs={"property": "og:description"})
    if meta:
        description = meta.get("content", "")

    logger.info("Crawled %s — %d images, %d emails", url, len(images), len(emails))
    return {
        "images": images,
        "emails": list(emails),
        "title": title,
        "description": description,
    }
